qa/acceptance-tests/src/test/java/io/camunda/it/client/RolesByTenantIntegrationTest.java

```diff
@@ -0,0 +1,242 @@
+    // when and then
+    assertThatThrownBy(
+            () -> camundaClient.newAssignRoleToTenantCommand(tenantId).roleId(roleId).send().join())
+        .isInstanceOf(ProblemException.class)
+        .hasMessageContaining("409: 'Conflict'")
+        .hasMessageContaining("the role is already assigned to the tenant");
+  }
+
+  @Test
+  void shouldReturnAllRolesByTenant() {
+    // given
+    final var tenantId = "tenant-" + Strings.newRandomValidIdentityId();
+    final var firstRoleId = "role-" + Strings.newRandomValidIdentityId();
+    final var secondRoleId = "role-" + Strings.newRandomValidIdentityId();
+    createTenant(tenantId);
+    createRole(firstRoleId, "name1", "desc1");
+    createRole(secondRoleId, "name2", "desc2");
+
+    camundaClient.newAssignRoleToTenantCommand(tenantId).roleId(firstRoleId).send().join();
+    camundaClient.newAssignRoleToTenantCommand(tenantId).roleId(secondRoleId).send().join();
+
+    // when and then
+    Awaitility.await("Roles should be visible in tenant role search")
+        .atMost(TIMEOUT_DATA_AVAILABILITY)
+        .untilAsserted(
+            () -> {
+              final var roles = camundaClient.newRolesByTenantSearchRequest(tenantId).send().join();
+              assertThat(roles.items())
+                  .extracting("roleId")
+                  .containsExactlyInAnyOrder(firstRoleId, secondRoleId);
+            });
+  }
+
+  @Test
+  void shouldReturnEmptyListForTenantWithoutRoles() {
+    final String emptyTenant = "empty-tenant-" + Strings.newRandomValidIdentityId();
+    createTenant(emptyTenant);
+
+    final var roles = camundaClient.newRolesByTenantSearchRequest(emptyTenant).send().join();
+
+    assertThat(roles.items()).isEmpty();
+  }
+
+  @Test
+  void shouldRejectSearchIfTenantIdIsNull() {
+    assertThatThrownBy(() -> camundaClient.newRolesByTenantSearchRequest(null).send().join())
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("tenantId must not be null");
+  }
+
+  @Test
+  void shouldRejectSearchIfTenantIdIsEmpty() {
+    assertThatThrownBy(() -> camundaClient.newRolesByTenantSearchRequest("").send().join())
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("tenantId must not be empty");
+  }
+
+  @Test
+  void shouldUnassignRoleFromTenant() {
+    // given
+    final var tenantId = "tenant-" + Strings.newRandomValidIdentityId();
+    final var roleId = "role-" + Strings.newRandomValidIdentityId();
+    createTenant(tenantId);
+    createRole(roleId, "name", "desc");
+    camundaClient.newAssignRoleToTenantCommand(tenantId).roleId(roleId).send().join();
+    // when
+    camundaClient.newUnassignRoleFromTenantCommand(tenantId).roleId(roleId).send().join();
+    // then
+    final var roles = camundaClient.newRolesByTenantSearchRequest(tenantId).send().join();
+    assertThat(roles.items()).noneMatch(role -> role.getRoleId().equals(roleId));
```

Comment:
🔧 This is one cause of flakiness. It needs the await as we are asserting a query to secondary storage.